shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java

```diff
@@ -110,18 +132,32 @@ public void startInitialization(@NonNull Context applicationContext, @NonNull Se
     }
 
     // Ensure that the context is actually the application context.
-    applicationContext = applicationContext.getApplicationContext();
+    final Context appContext = applicationContext.getApplicationContext();
 
     this.settings = settings;
 
     initStartTimestampMillis = SystemClock.uptimeMillis();
-    initResources(applicationContext);
 
-    System.loadLibrary("flutter");
-
-    VsyncWaiter.getInstance(
-            (WindowManager) applicationContext.getSystemService(Context.WINDOW_SERVICE))
+    VsyncWaiter.getInstance((WindowManager) appContext.getSystemService(Context.WINDOW_SERVICE))
         .init();
+
+    // Use a background thread for initialization tasks that require disk access.
+    Callable<InitResult> initTask =
+        new Callable<InitResult>() {
+          @Override
+          public InitResult call() {
+            ResourceExtractor resourceExtractor = initResources(appContext);
+
+            System.loadLibrary("flutter");
+
+            return new InitResult(
```

Comment:
love that we finally stopped using the resource extractor as a unit of transaction now since it was really confusing. Though this is close to getting us all the way. i.e. upon initialization task complete, initialization isn't quite done yet. We still need to wait for the resource extractor to finish. Perhaps just chain the resource extractor completion inside this async task?

Reply:
done